### DISCUSSION TITLE
Consistent ~2% decrease in HTTP response times with 3.5.0-RC3

### DISCUSSION THREAD

**Discussion by gvolpe** [typelevel/cats-effect#3502] [Q&A] [answered]
Hi folks 👋🏽 ,

We're using the Typelevel stack at work on a greenfield project, so I took some time to run some [Gatling](https://gatling.io/) load tests we have to compare the HTTP response times performance using the latest RC. We also use Http4s with the Ember server implementation. 

Here's the version matrix I've managed to test, all with **Scala 3.2.2**:

| Cats Effect | Http4s |
|--------------|--------------|
| 3.4.8 | 0.23.18 |
| 3.5.0-RC3 | 0.23.19-RC3 |
| 3.5.0-RC3 | 0.23.18 |

We run our service on AWS and the main database is DynamoDB. Though, all these tests were run on my machine against Localstack.

The load test performs multiple CRUD operations regarding a user profile (simply a User ID and a name for now). 

---

# Low input values

| Name | Value |
|---------|----------|
| rampUpTime | 1|           
| rampFromUsersPerSec | 1 |
| rampToUsersPerSec | 1 |
| constantUsersPerSec | 1 |
| constantTime | 1 |

## Cats Effect 3.4.8 / Http4s 0.23.18

Response time **P99: 126 ms** / **Max: 177 ms**

![ce-3 4-simple](https://user-images.githubusercontent.com/443978/226381535-72d5d488-4026-4162-992c-dd41fa051c74.png)

## Cats Effect 3.5.0-RC3 / Http4s 0.23.19-RC3

Response time **P99: 137 ms** / **Max: 607 ms**

![ce-3 5-rc-simple](https://user-images.githubusercontent.com/443978/226381474-21bd799c-4494-42fb-894d-b9274489a5f7.png)

## Cats Effect 3.5.0-RC3 / Http4s 0.23.18

I've run this one only to ensure changes in Http4s didn't affect the results, and it seems consistent across multiple runs.

Response time **P99: 136 ms** / **Max: 1682 ms**

![ce-3 4-http4s-0 23 18-simple](https://user-images.githubusercontent.com/443978/226381425-197e85d8-22bb-4fad-8a93-8376087e366b.png)

---

# High input values

| Name | Value |
|---------|----------|
| rampUpTime | 1|           
| rampFromUsersPerSec | 1 |
| rampToUsersPerSec | 8 |
| constantUsersPerSec | 8 |
| constantTime | 1 |

## Cats Effect 3.4.8 / Http4s 0.23.18

Response time **P99: 18.848 seconds** / **Max: 19.850 seconds**

![ce-3 4-complex](https://user-images.githubusercontent.com/443978/226381344-d4dae6e1-7221-434b-9847-41c3788e754e.png)

## Cats Effect 3.5.0-RC3 / Http4s 0.23.19-RC3

Response time **P99: 19.247 seconds** / **Max: 20.077 seconds**

![ce-3 5-rc-complex](https://user-images.githubusercontent.com/443978/226381192-7f13554f-c54d-4c24-9824-e56abaef1c00.png)

---

# Summary

Multiple runs using different input values seem consistent, showing a **~2% decrease** in response times. However, the main bottleneck in our case is DynamoDB I/O, so we have bigger issues to solve 😄 (luckily this is not PROD yet), but I thought it was worth reporting.

Now I realize this may not be of great help, given that the code belongs to the company I work for, so I can't share implementation details. Though, I'd be happy to run further tests if you think I could try other settings / versions, etc.

**Comment by djspiewak** [comment, marked as answer]
This is *immensely* helpful thank you! This is in line with @wjoel's results and suggests that my `hey` benchmarks are genuinely not capturing a representative scenario.

I think at a minimum we need to rework the timers data structure and fix the busy-loop scenario.

**Comment by djspiewak** [comment]
Following up on this: any chance you would be willing to take a look at `RC5`? Based on the testing and benchmarking we've done, this should more than resolve the regressions you've seen, and ideally leave you in a much better place than 3.4.x.

**Reply by gvolpe** [reply]
That's awesome to hear, will definitely give it a shot on Monday and report back, appreciate all the work you guys do 🙇🏽

**Comment by gvolpe** [comment]
# Cats Effect 3.5.0-RC5

Following up with testing the latest RC5. Here's the version matrix I've managed to test, all with **Scala 3.2.2**:

| Cats Effect | Http4s |
|--------------|--------------|
| 3.4.9 | 0.23.18 |
| 3.5.0-RC5 | 0.23.19-RC3 |
| 3.5.0-RC5 | 0.23.18 |

Once again, all these tests were run on my machine against Localstack (AWS platform).

The load test performs multiple CRUD operations regarding a user profile (simply a User ID and a name for now). 

---

# Low input values

| Name | Value |
|---------|----------|
| rampUpTime | 1|           
| rampFromUsersPerSec | 1 |
| rampToUsersPerSec | 1 |
| constantUsersPerSec | 1 |
| constantTime | 1 |

## Cats Effect 3.4.9 / Http4s 0.23.18

Response time **P99: 145 ms** / **Max: 264 ms**

![ce-3 4 9-simple](https://user-images.githubusercontent.com/443978/235711194-b51dd44f-5d05-4c2e-a28f-4a50be0e2d71.png)

## Cats Effect 3.5.0-RC5 / Http4s 0.23.19-RC3

Response time **P99: 161 ms** / **Max: 1727 ms**

![ce-3 5-rc5-simple](https://user-images.githubusercontent.com/443978/235711313-c0761f71-df0b-4869-8146-62b0ccce5863.png)

## Cats Effect 3.5.0-RC5 / Http4s 0.23.18

Response time **P99: 123 ms** / **Max: 161 ms**

![ce-3 5-rc5-http4s-0 23 18-simple](https://user-images.githubusercontent.com/443978/235711635-142ee4bd-4f38-4f79-afa3-d9d674077bd0.png)

---

# High input values

| Name | Value |
|---------|----------|
| rampUpTime | 1|           
| rampFromUsersPerSec | 1 |
| rampToUsersPerSec | 8 |
| constantUsersPerSec | 8 |
| constantTime | 1 |

## Cats Effect 3.4.9 / Http4s 0.23.18

Response time **P99: 18.151 seconds** / **Max: 21.639 seconds**

![ce-3 4 9-complex](https://user-images.githubusercontent.com/443978/235712413-e127cea5-101c-4c66-9b08-e6ebeda5b0d1.png)

## Cats Effect 3.5.0-RC5 / Http4s 0.23.19-RC3

Response time **P99: 20.260 seconds** / **Max: 20.915 seconds**

![ce-3 5-rc5-complex](https://user-images.githubusercontent.com/443978/235715480-9b3dea21-01ef-40c7-8632-28bb92767b90.png)

## Cats Effect 3.5.0-RC5 / Http4s 0.23.18

Two runs with the same arguments.

1. Response time **P99: 19.817 seconds** / **Max: 20.795 seconds**
2. Response time **P99: 21.163 seconds** / **Max: 21.969 seconds**

![ce-3 5-rc5-http4s-0 23 18-simple](https://user-images.githubusercontent.com/443978/235715645-9a2bbef6-5213-4fc3-89b9-b44e6ae0edcb.png)

---

# Summary

There still seem to be a slight decrease in response times, even though one of the tests with low input values gave better results. Didn't have much more time to keep analyzing this (busy times at work!), but I could try to follow up over the weekend.

**Reply by djspiewak** [reply]
Unless I'm misinterpreting the results, it looks like 3.5.0 actually brings down the response times a bit (consistent with what @wjoel saw in his testing), but http4s 0.23.19-RC3 gives back those gains and then some. No? The ramp test seems to have too much variance for me to trust it, but I tend to put a lot of weight on the p99s.

**Reply by gvolpe** [reply]
Sorry about that rushed up post, had limited time and wasn't enough to make much sense out of it. I've run them again 3 times each for fairness, as the first one always seems to take a slight hit on warm up. 

### P99 response times for low input values

| Cats Effect | Http4s | P99 L1 | P99 L2 | P99 L3 |
|--------------|--------------|--------|------------|-------------|
| 3.4.9 | 0.23.18 | 146 ms | 122 ms | 126 ms |
| 3.5.0-RC5 | 0.23.18 | 133 ms | 127 ms | 120 ms |
| 3.5.0-RC5 | 0.23.19-RC3 | 142 ms | 134 ms | 141 ms |

### P99 response times for high input values 

| Cats Effect | Http4s | P99 H1 | P99 H2 | P99 H3 |
|--------------|--------------|--------|------------|-------------|
| 3.4.9 | 0.23.18 | 19.462 s | 20.298 s | 20.141 s |
| 3.5.0-RC5 | 0.23.18 | 17.690 s | 18.213 s | 17.079 s |
| 3.5.0-RC5 | 0.23.19-RC3 | 20.626 s | 22.016 s | 22.884 s |

Cats Effect 3.5.0-RC5 seems to perform a bit better than v3.4.9 when using the same Http4s version. Though, Http4s 0.23.19-RC3 seems to perform worse on any Cats Effect version. Hope that helps!

**Reply by armanbilge** [reply]
@gvolpe it's a lot to ask (this has been super helpful so far, thank you!), but would you be able to try http4s 0.23.19 RC1 and RC2 as well? I'm hoping we can narrow down when exactly the performance regression happened and hopefully do something about it.

**Reply by gvolpe** [reply]
@armanbilge not a problem, I'll try to make time for it, because these tests take ages to run! 😄

**Reply by djspiewak** [reply]
In the meantime though, I think this (taken together with @wjoel's results) is the golden signal to move forward with 3.5 from a performance standpoint.

**Reply by gvolpe** [reply]
@armanbilge the regression seems to have happened on RC3. Here's the updated results:

### P99 response times for low input values

| Cats Effect | Http4s | P99 L1 | P99 L2 | P99 L3 |
|--------------|--------------|--------|------------|-------------|
| 3.4.9 | 0.23.18 | 146 ms | 122 ms | 126 ms |
| 3.5.0-RC5 | 0.23.18 | 133 ms | 127 ms | 120 ms |
| 3.5.0-RC5 | 0.23.19-RC3 | 142 ms | 134 ms | 141 ms |
| 3.5.0-RC5 | 0.23.19-RC2 | 147 ms | 126 ms | 128 ms |

### P99 response times for high input values 

| Cats Effect | Http4s | P99 H1 | P99 H2 | P99 H3 |
|--------------|--------------|--------|------------|-------------|
| 3.4.9 | 0.23.18 | 19.462 s | 20.298 s | 20.141 s |
| 3.5.0-RC5 | 0.23.18 | 17.690 s | 18.213 s | 17.079 s |
| 3.5.0-RC5 | 0.23.19-RC3 | 20.626 s | 22.016 s | 22.884 s |
| 3.5.0-RC5 | 0.23.19-RC2 | 19.868 s | 19.949 s | 17.768 s |

I didn't even test Http4s 0.23.19-RC1 as it seemed redundant :)

**Reply by armanbilge** [reply]
Thanks! From those numbers it seems that even RC2 is a bit slower, but indeed most of the regression seems to be RC3. Hum, what could it be 🤔
